### PR TITLE
introduce cross-platform automatic builds

### PR DESCRIPTION
**Pull request by holisticode**
This PR introduces the ability to run cross-platform builds automatically.

This current configuration triggers a new build after **a new tag has been created**.

Therefore, the way to trigger builds is to:
* Create a tag: `tag -a <TAG> -m "<TEXT>"`, e.g. `tag -a .0.6.0 -m "everest"`
* Push a tag to the repository: `git push origin tags/<TAG>`, e.g. `git push origin tags/0.6.0`.

The build then will trigger each:
* A `linux` build in `deb` (debian package) and `tgz` (tar.gz) format
* A `windows` build (Windows server 2019)
* A `macos` build (Catalina)

All the artifacts are then pushed to storage.

The trigger (tags) could be changed. Triggers can be:
* push
* pull_request
* release (new release created)

There should be a security review of this PR.

### CODE REVIEW

**Review by aaronbuchwald — DISMISSED** [review]

File: .github/workflows/build-deb-pkg.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
+PKG_ROOT=/tmp/avalanche
+DEBIAN_BASE_DIR=$PKG_ROOT/debian
+AVALANCHE_BUILD_BIN_DIR=$DEBIAN_BASE_DIR/usr/local/bin
+AVALANCHE_LIB_DIR=$DEBIAN_BASE_DIR/usr/local/lib/avalanche
+TEMPLATE=.github/workflows/debian/template 
+DEBIAN_CONF=$DEBIAN_BASE_DIR/DEBIAN
+
+mkdir -p $PKG_ROOT/$DEBIAN_BASE_DIR
```
Comment (moreati) [inline .github/workflows/build-deb-pkg.sh:8]: Won't this create `/tmp/avalanche/tmp/avalanche/debian`? $DEBIAN_BASE_DIR already includes $PKG_ROOT

File: .github/workflows/build-deb-pkg.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+PKG_ROOT=/tmp/avalanche
+DEBIAN_BASE_DIR=$PKG_ROOT/debian
+AVALANCHE_BUILD_BIN_DIR=$DEBIAN_BASE_DIR/usr/local/bin
+AVALANCHE_LIB_DIR=$DEBIAN_BASE_DIR/usr/local/lib/avalanche
```
Comment (moreati) [inline .github/workflows/build-deb-pkg.sh:4]: Would the final path for the evm plugin be `/usr/local/lib/avalanche/evm`, or `/usr/local/lib/avalanche/plugins/evm`?

I went for `/usr/local/lib/avalanche/plugins/evm` with the Ansible roles in `scripts/ansible`, but I'm open to changing that.

**Comment by moreati** [comment]
Reminders to myself, these are not objections or comments
 - The .deb includes gecko binary, and plugins only. User(s), and service(s) are not.
 - The .deb doesn't declare any dependencies. Gecko is a Go binary, hence statically linked, so I think this is correct.

File: .github/workflows/build-deb-pkg.sh (hunk 1 above)
Comment (holisticode) [reply .github/workflows/build-deb-pkg.sh:8]: you are right of course

File: .github/workflows/build-deb-pkg.sh (hunk 2 above)
Comment (holisticode) [reply .github/workflows/build-deb-pkg.sh:4]: The final dir is `/usr/local/lib/avalanche/evm` . I am still pushing though for automatic lookup of the gecko client to find it there though @StephenButtolph . So far we had `/usr/local/lib/gecko/evm`, but after the renaming of the binary I did proactively change it to that.

**Comment by holisticode** [comment]
@moreati your comments are correct.

Regarding users/services: We haven't included users/systemd services. `geth` doesn't do it either. It's easier for sure. The main reason is that we would have to add a script to create a user, and it might not be immediately visible to users where to find the data.

All fixable things but not highest prio right now.

File: .github/workflows/build-deb-pkg.sh (hunk 2 above)
Comment (moreati) [reply .github/workflows/build-deb-pkg.sh:4]: /usr/local/lib/avalanche/**plugins**/evm makes slightly more sense to me. It namespaces things a bit, making it easier to distinguish other potential extension mechanisms or hooks in the future.

Comment (holisticode) [reply .github/workflows/build-deb-pkg.sh:4]: I am not sure I agree.

The binary itself is somewhere else in `/usr/local/bin/` - we don't have anything other than plugins so far, and `/usr/local/lib/avalanche` is already like the `plugins` folder in some sense. Unless I am missing something about upcoming potential other binaries which could be placed there. 

I am also moving away from my old tendency to want to organize code for things in the future which are not a need today.

Let's see if someone else chimes in and offers an opinion.

File: .github/workflows/build-deb-pkg.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+TEMPLATE=.github/workflows/debian/template 
+DEBIAN_CONF=$DEBIAN_BASE_DIR/DEBIAN
+
+mkdir -p $DEBIAN_BASE_DIR
+mkdir -p $DEBIAN_CONF
+mkdir -p $AVALANCHE_BUILD_BIN_DIR
+mkdir -p $AVALANCHE_LIB_DIR
+
+OK=`cp ./build/avalanche $AVALANCHE_BUILD_BIN_DIR`
```
Comment (moreati) [inline .github/workflows/build-deb-pkg.sh:13]: The binary formerly known as `gecko`, and `avalanche` is now called `avalanchego`. This (and probably other parts) will need updating to reflect that.

File: .github/workflows/build-deb-pkg.sh (hunk 2 above)
Comment (moreati) [inline .github/workflows/build-deb-pkg.sh:4]: For the Ansible roles I'm currently debating whether the directories should remain `/usr/local/lib/avalanche/...`, or change to `/usr/local/lib/avalanchego/...`. What do you think?

Independent of whether `'plugins` is part of the path.

File: .github/workflows/debian/template/control [hunk 4]
```diff
@@ -0,0 +1,8 @@
+Package: avalanche
```
Comment (moreati) [inline .github/workflows/debian/template/control:1]: This should probably change to avalanchego.

File: .github/workflows/debian/template/control [hunk 5]
```diff
@@ -0,0 +1,8 @@
+Package: avalanche 
+Version: 0.1.0
```
Comment (moreati) [inline .github/workflows/debian/template/control:2]: Debian package version is hard coded.

File: .github/workflows/build-deb-pkg.sh (hunk 2 above)
Comment (moreati) [reply .github/workflows/build-deb-pkg.sh:4]: Given that `~/.gecko/` has also been renamed to `~/.avalanchego`, I'm leaning toward using `/usr/local/lib/avalanchego/...`

Comment (holisticode) [reply .github/workflows/build-deb-pkg.sh:4]: Internally we chose `/usr/local/lib/avalanchego/`

File: .github/workflows/debian/template/control (hunk 5 above)
Comment (holisticode) [reply .github/workflows/debian/template/control:2]: It is only a place holder. The `build-deb-pkg.sh` script replaces the content of the placeholder with the current version.

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**